pkg/agent/client/client.go

```diff
@@ -0,0 +1,213 @@
+}
+
+// Config holds a client configuration
+type Config struct {
+	Addr        net.Addr
+	Log         logrus.FieldLogger
+	TrustDomain url.URL
+	// KeysAndBundle is a callback that must return the keys and bundle used by the client
+	// to connect via mTLS to Addr.
+	KeysAndBundle func() (*x509.Certificate, *ecdsa.PrivateKey, []*x509.Certificate)
+}
+
+type Update struct {
+	Entries map[string]*common.RegistrationEntry
+	SVIDs   map[string]*node.Svid
+	Bundle  []byte
+}
+
+type client struct {
+	c    *Config
+	conn *grpc.ClientConn
+	m    sync.Mutex
+	// Callback to be used for testing purposes.
+	newNodeClientCallback func() (node.NodeClient, error)
+}
+
+// New creates a new client struct with the configuration provided
+func New(c *Config) *client {
+	return &client{
+		c: c,
+	}
+}
+
+func newGRPCConn(c *Config) (*grpc.ClientConn, error) {
+	credFunc := func() (credentials.TransportCredentials, error) {
+		var tlsCert []tls.Certificate
+		var tlsConfig *tls.Config
+
+		svid, key, bundle := c.KeysAndBundle()
+		spiffePeer := &spiffe_tls.TLSPeer{
+			SpiffeIDs:  []string{"spiffe://" + c.TrustDomain.Host + "/spiffe/server"},
+			TrustRoots: util.NewCertPool(bundle...),
+		}
+		tlsCert = append(tlsCert, tls.Certificate{Certificate: [][]byte{svid.Raw}, PrivateKey: key})
+		tlsConfig = spiffePeer.NewTLSConfig(tlsCert)
+		return credentials.NewTLS(tlsConfig), nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second) // TODO: Make this timeout configurable?
+	defer cancel()
+
+	config := grpcutil.GRPCDialerConfig{
+		Log:      grpcutil.LoggerFromFieldLogger(c.Log),
+		CredFunc: credFunc,
+	}
+	dialer := grpcutil.NewGRPCDialer(config)
+	conn, err := dialer.Dial(ctx, c.Addr)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create connection: %v", err)
+	}
+	return conn, nil
+}
+
+func (c *client) FetchUpdates(req *node.FetchSVIDRequest) (*Update, error) {
+	nodeClient, err := c.newNodeClient()
+	if err != nil {
+		return nil, err
+	}
+
+	stream, err := nodeClient.FetchSVID(context.Background())
+	// We weren't able to get a stream...close the client and return the error.
+	if err != nil {
+		c.Release()
+		c.c.Log.Errorf("%v: %v", ErrUnableToGetStream, err)
+		return nil, ErrUnableToGetStream
+	}
+
+	// Send the request to the server using the stream.
+	err = stream.Send(req)
+	// Close the stream whether there was an error or not
+	stream.CloseSend()
+	if err != nil {
+		// TODO: should we try to create a new stream?
+		return nil, err
+	}
+
+	regEntries := map[string]*common.RegistrationEntry{}
+	svids := map[string]*node.Svid{}
+	var lastBundle []byte
+	// Read all the server responses from the stream.
+	for {
+		resp, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			if len(regEntries) > 0 || len(svids) > 0 || lastBundle != nil {
+				// There was an error receiving a response, exit loop to return what we have.
+				return &Update{regEntries, svids, lastBundle}, ErrPartialResponse
```

Comment:
Seems like returning `ErrPartialResponse` will obscure the original error from the caller? Perhaps it would be better to always build the Update struct response, and return the original error. If the caller wants to still try to use the Update even after receiving non-nil error, they can do the above checks themselves.